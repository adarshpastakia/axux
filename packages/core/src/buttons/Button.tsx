/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

import { handleClick } from "@axux/utilities/src/handlers";
import { forwardRef, useMemo, type FC } from "react";
import { AxAnimation } from "../animations";
import { Link, type LinkProps } from "../components/Link";
import { useBadge } from "../hooks/useBadge";
import { getTooltipProps } from "../hooks/useTooltip";
import { AxHotKey } from "../hotkeys/HotKey";
import { AxIcon } from "../icons/Icon";
import {
  type BadgeType,
  type ChildrenProp,
  type Color,
  type ElementProps,
  type IconProp,
  type MouseProps,
  type RefProp,
  type Size,
  type TooltipType,
} from "../types";
import { AppIcons } from "../types/appIcons";
import { ActionButton } from "./Action";
import { ConfirmButton } from "./Confirm";
import { DropdownButton } from "./Dropdown";

export interface ButtonProps
  extends RefProp,
    IconProp,
    MouseProps,
    ElementProps,
    Partial<ChildrenProp> {
  /**
   * size
   */
  size?: Size;
  /**
   * color
   */
  color?: Color;
  /**
   * button style
   */
  variant?: "outline" | "solid" | "link" | "normal";
  /**
   * icon alignment
   */
  iconAlign?: "start" | "end" | "top";
  /**
   * button type
   */
  type?: "button" | "submit" | "reset";
  /**
   * full width
   */
  fullWidth?: boolean;
  /**
   * active state
   */
  isActive?: boolean;
  /**
   * rounded button
   */
  isRound?: boolean;
  /**
   * loading state
   */
  isLoading?: boolean;
  /**
   * disabled state
   */
  isDisabled?: boolean;
  /**
   * hide dropdown caret
   */
  hideCaret?: boolean;
  /**
   * use loading spinner
   */
  useSpinner?: boolean;
  /**
   * disable tab focus
   */
  noTabFocus?: boolean;
  /**
   * tooltip
   */
  tooltip?: TooltipType;
  /**
   * badge
   */
  badge?: BadgeType;
  /**
   * hot key
   */
  hotKey?: string;
  /**
   * navigation props
   */
  nav?: LinkProps;
}

export interface ButtonGroupProps extends ElementProps, ChildrenProp {
  isVertical?: boolean;
  variant?: "normal" | "flat" | "plain";
}

/**
 * Button group
 * @prop className
 * @prop isVertical
 * @prop variant - normal | flat | plain
 */
export const ButtonGroup: FC<ButtonGroupProps> = ({
  children,
  variant = "normal",
  isVertical = false,
  className = "",
  ...rest
}: ButtonGroupProps) => (
  <div
    className={`ax-button__group ${className ?? ""}`}
    data-variant={variant}
    data-vertical={isVertical}
    {...rest}
  >
    {children}
  </div>
);

/**
 * A simple clickable element used to trigger a specific action or function in the user interface
 */
export const ButtonComponent: FC<ButtonProps> = forwardRef<
  HTMLElement,
  ButtonProps
>((props, ref) => {
  const {
    icon,
    variant: style,
    size,
    children,
    type = "button",
    color,
    rtlFlip,
    noTabFocus,
    iconAlign,
    className,
    isActive = false,
    isRound = false,
    isLoading = false,
    isDisabled = false,
    hideCaret = true,
    tooltip,
    badge,
    fullWidth,
    hotKey,
    onClick,
    useSpinner,
    stopPropagation = false,
    // @ts-expect-error ignore
    "data-extra": extra,
    // @ts-expect-error ignore
    "data-tool": tool,
    // @ts-expect-error ignore
    "data-popover-open": popoverOpen,
    ...rest
  } = props;

  const Badge = useBadge(badge);
  const isSquare = useMemo(() => !!icon && !children, [icon, children]);

  const tooltipProps = useMemo(() => getTooltipProps(tooltip), [tooltip]);

  const clickHandler = useMemo(() => {
    if (!isDisabled && !isLoading && !isActive) {
      return handleClick(onClick, { stopPropagation });
    }
  }, [onClick, isActive, isDisabled, isLoading, stopPropagation]);

  /** ***************** component *******************/
  return (
    <div
      data-type={style}
      data-size={size}
      data-tool={tool}
      data-color={color}
      data-active={isActive}
      data-round={isRound}
      data-block={fullWidth}
      data-square={isSquare}
      data-spinner={!!useSpinner}
      data-loading={isLoading}
      data-disabled={isDisabled}
      data-icon-align={iconAlign}
      className={`ax-button ${className ?? ""}`}
    >
      {hotKey && <AxHotKey keyCombo={hotKey} handler={onClick} />}
      {!useSpinner && isLoading && <div className="ax-button__loader" />}
      <Link
        {...rest}
        type={type}
        as="button"
        role="button"
        ref={ref}
        className="ax-button__inner"
        onClick={clickHandler}
        data-disabled={isDisabled || isLoading}
        data-popover-open={popoverOpen}
        {...tooltipProps}
        tabIndex={noTabFocus ? -1 : 0}
      >
        {!(useSpinner && isLoading) && icon && (
          <AxIcon className="ax-button__icon" rtlFlip={rtlFlip} icon={icon} />
        )}
        {useSpinner && isLoading && (
          <AxAnimation.Spinner className="ax-button__icon" />
        )}
        {children && <div className="ax-button__label">{children}</div>}
        {Badge}
        {hotKey && <AxHotKey.Label keyCombo={hotKey} />}
        {!hideCaret && (
          <AxIcon className="ax-button__caret" icon={AppIcons.iconCaretDown} />
        )}
      </Link>
      {extra}
    </div>
  );
});
ButtonComponent.displayName = "AxButton";

export const AxButton = Object.assign(ButtonComponent, {
  Group: ButtonGroup,
  Action: ActionButton,
  Confirm: ConfirmButton,
  Dropdown: DropdownButton,
});

AxButton.displayName = "AxButton";
AxButton.Action.displayName = "AxButton.Action";
AxButton.Confirm.displayName = "AxButton.Confirm";
AxButton.Dropdown.displayName = "AxButton.Dropdown";
AxButton.Group.displayName = "AxButton.Group";
