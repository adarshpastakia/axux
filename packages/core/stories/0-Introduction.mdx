import { Title, Canvas, Story } from "@storybook/addon-docs";
import { Anchor, Subheading } from "@storybook/blocks";
import { Meta } from "@storybook/blocks";

<Meta title="@core/Introduction" />

# Core Components

<Anchor storyId="install">
  <Subheading>Installing</Subheading>
</Anchor>

```bash

#npm
npm install @axux/core

# yarn
yarn add @axux/core

```

<Anchor storyId="using">
  <Subheading>Using AxUX</Subheading>
</Anchor>

```tsx
import { AxApplicationProvider,AxViewport } from "@axux/core";
import { I18nextProvider } from "react-i18next";
import { BrowserRouter } from "react-router";

export const Application = () => {
  return (
    {/** i18n for localization **/}
    <I18nextProvider>
      {/** react router **/}
      <BrowserRouter>
        <AxApplicationProvider>
          <AxViewport>
            .
            .
            .
            .
          </AxViewport>
        </AxApplicationProvider>
      </BrowserRouter>
    </I18nextProvider>
  );
};
```
