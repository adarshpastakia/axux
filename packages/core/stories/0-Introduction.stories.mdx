import { Title, Canvas, Story } from "@storybook/addon-docs";

<Meta title="@core/Introduction" />

<Title>Core Components</Title>
