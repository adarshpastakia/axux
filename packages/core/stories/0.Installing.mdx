import { Title, Canvas, Story } from "@storybook/addon-docs";
import { Anchor, Subheading } from "@storybook/blocks";
import { Meta } from "@storybook/blocks";

<Meta title="@core/Installing" />

# Core Components

<Anchor storyId="install">
  <Subheading>Installing</Subheading>
</Anchor>

```bash

#npm
npm install @axux/core

# yarn
yarn add @axux/core

```

<Anchor storyId="using">
  <Subheading>Using AxUX</Subheading>
</Anchor>

<cite type="code">src/index.tsx</cite>

```tsx
import { AxApplicationProvider, AxViewport } from "@axux/core";
import { Fragment, StrictMode } from "react";
import ReactDOM from "react-dom/client";
import { I18nextProvider } from "react-i18next";
import { Outlet, RouterProvider, createMemoryRouter } from "react-router";
import i18next from "./i18n";
import reportWebVitals from "./reportWebVitals";
import "./styles.css";

const root = ReactDOM.createRoot(document.getElementById("root"));

const router = createMemoryRouter([
  {
    path: "/",
    element: (
      <AxViewport>
        <div>Build the ui here</div>
        <Outlet />
      </AxViewport>
    ),
    children: [
      {
        path: "page",
        element: <Page />,
      },
      {
        path: "page",
        element: <Page />,
      },
    ],
  },
]);

const Wrapper = process.env.NODE_ENV === "development" ? Fragment : StrictMode;
root.render(
  <Wrapper>
    <AxApplicationProvider defaultPrimary="denim" defaultAccent="coral">
      <I18nextProvider i18n={i18next}>
        <RouterProvider router={router} />
      </I18nextProvider>
    </AxApplicationProvider>
  </Wrapper>,
);

// @ts-expect-error ignore
if (module.hot) {
  // @ts-expect-error ignore
  module.hot.accept();
}

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals(console.debug);
```
