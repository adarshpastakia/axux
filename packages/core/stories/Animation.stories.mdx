import {
  Anchor,
  ArgsTable,
  Canvas,
  Description,
  Heading,
  Meta,
  Story,
  Title,
} from "@storybook/addon-docs";
import { useState, useEffect } from "react";
import { AxAnimation, AxBadge, AxProgress } from "../src";

<Meta title="@core/Animations" />

<Title>Animations</Title>

<Heading storyId="core-animations--1a-check">Animated Circles</Heading>

<Canvas>
  <Story name="1a.Check">
    <AxAnimation.Check className="text-success-500" />
  </Story>
  <Story name="1b.Cross">
    <AxAnimation.Cross className="text-danger-500" />
  </Story>
  <Story name="1c.Info">
    <AxAnimation.Info className="text-primary-500" />
  </Story>
</Canvas>

<ArgsTable sort="requiredFirst" of={AxAnimation.Check} />

<Heading storyId="core-animations--2-a-bars">Animated Indicators</Heading>

<Canvas>
  <Story name="2a.Bars">
    <AxAnimation.Bars className="text-primary-500" />
  </Story>
  <Story name="2b.Spinner">
    <AxAnimation.Spinner className="text-4xl text-accent-500" />
  </Story>
  <Story name="2c.Card">
    <AxAnimation.Card className="w-64" showIcon />
  </Story>
</Canvas>

<ArgsTable sort="requiredFirst" of={AxAnimation.Bars} />

<Heading storyId="core-animations--3-a-progress-circle">Progress</Heading>

export const CircleTemplate = (args) => {
  const [value, setValue] = useState(24);
  useEffect(() => {
    const tmr = setInterval(() => {
      setValue(Math.random() * 100);
    }, 5000);
    return () => {
      clearInterval(tmr);
    };
  }, []);
  return <AxProgress.Circle size="md" value={value} />;
};
const CircleStory = CircleTemplate.bind({});
CircleStory.args = {};

export const BarTemplate = (args) => {
  const [value, setValue] = useState(24);
  useEffect(() => {
    const tmr = setInterval(() => {
      setValue(Math.random() * 100);
    }, 5000);
    return () => {
      clearInterval(tmr);
    };
  }, []);
  return (
    <div style={{ width: "40rem" }}>
      <AxProgress.Bar size="md" value={value} animate />
    </div>
  );
};
const BarStory = BarTemplate.bind({});
BarStory.args = {};

<Canvas>
  <Story name="3a.Progress Circle">{CircleStory}</Story>
  <Story name="3b.Progress Bar">{BarStory}</Story>
</Canvas>

<ArgsTable sort="requiredFirst" of={AxProgress.Bar} />
