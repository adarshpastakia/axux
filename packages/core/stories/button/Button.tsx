/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */

import { action } from "@storybook/addon-actions";
import { ComponentStory } from "@storybook/react";
import {
  AxApplicationProvider,
  AxButton,
  AxContent,
  AxFlexBox,
  AxHeader,
  AxPanel,
  AxTitle,
} from "../../src";
import { HotKeyWrapper } from "../../src/hotkeys/HotKeyWrapper";
import { AxMenu } from "../../src/menu/Menu";

export const ButtonTemplate: ComponentStory<typeof AxButton> = (props) => (
  <AxButton {...props} />
);

export const ButtonStory = ButtonTemplate.bind({});
ButtonStory.args = {
  children: "Click Me!",
};

export const OnlyIconStory = ButtonTemplate.bind({});
OnlyIconStory.args = {
  icon: "mdi mdi-check-circle",
};
export const WithIconStory = ButtonTemplate.bind({});
WithIconStory.args = {
  icon: "mdi mdi-check-circle",
  children: "Click Me!",
};

export const ButtonSizes: ComponentStory<AnyObject> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row align="center">
      <AxFlexBox.Col>
        <AxButton {...props} size="sm">
          Small Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} size={undefined}>
          Normal Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} size="md">
          Medium Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} size="lg">
          Large Button!
        </AxButton>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonStyles: ComponentStory<AnyObject> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row align="center">
      <AxFlexBox.Col>
        <AxButton {...props} variant={undefined}>
          Normal Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} isRound>
          Rounded Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} variant="outline">
          Outline Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} variant="solid">
          Solid Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} variant="link">
          Link Button!
        </AxButton>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonStates: ComponentStory<AnyObject> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row align="center">
      <AxFlexBox.Col>
        <AxButton {...props} isDisabled>
          Disabled State!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} isLoading>
          Busy State!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} isLoading useSpinner>
          Busy State!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} isActive>
          Active State!
        </AxButton>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonColors: ComponentStory<AnyObject> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row align="center">
      <AxFlexBox.Col>
        <AxButton {...props} color="primary">
          Normal Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} color="accent">
          Accent Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} color="danger">
          Danger Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} color="success">
          Success Button!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton {...props} color="warning">
          Warning Button!
        </AxButton>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonTooltip: ComponentStory<AnyObject> = (props) => (
  <AxApplicationProvider>
    <AxFlexBox>
      <AxFlexBox.Row>
        <AxFlexBox.Col>
          <AxButton {...props} color={undefined} tooltip="Normal tooltip">
            Normal Tooltip!
          </AxButton>
        </AxFlexBox.Col>
        <AxFlexBox.Col>
          <AxButton
            {...props}
            color="primary"
            tooltip={{
              content: "Information tooltip",
              color: "info",
              placement: "top",
            }}
          >
            Info Tooltip!
          </AxButton>
        </AxFlexBox.Col>
        <AxFlexBox.Col>
          <AxButton
            {...props}
            color="danger"
            tooltip={{
              content: "Danger tooltip",
              color: "danger",
              placement: "right",
            }}
          >
            Danger Tooltip!
          </AxButton>
        </AxFlexBox.Col>
        <AxFlexBox.Col>
          <AxButton.Group>
            <AxButton tooltip="First">First</AxButton>
            <AxButton tooltip="Second">Second</AxButton>
            <AxButton tooltip="Third">Third</AxButton>
          </AxButton.Group>
        </AxFlexBox.Col>
      </AxFlexBox.Row>
    </AxFlexBox>
  </AxApplicationProvider>
);

export const ButtonBadges: ComponentStory<AnyObject> = (props) => (
  <HotKeyWrapper>
    <AxFlexBox>
      <AxFlexBox.Row>
        <AxFlexBox.Col>
          <AxButton {...props} badge={9}>
            Value Badge!
          </AxButton>
        </AxFlexBox.Col>
        <AxFlexBox.Col>
          <AxButton {...props} badge={{ ping: true, color: "success" }}>
            Pinging Badge!
          </AxButton>
        </AxFlexBox.Col>
        <AxFlexBox.Col>
          <AxButton
            {...props}
            badge={{
              value: "new",
              color: "primary",
              placement: "top",
            }}
          >
            Color Badge
          </AxButton>
        </AxFlexBox.Col>
        <AxFlexBox.Col>
          <AxButton
            {...props}
            badge={{
              ping: true,
              icon: "mdi mdi-bell",
              color: "danger",
              placement: "top-end",
            }}
          >
            Pinging Badge w/ Icon
          </AxButton>
        </AxFlexBox.Col>
        <AxFlexBox.Col>
          <AxButton {...props} hotKey="shift+?">
            Hot key
          </AxButton>
        </AxFlexBox.Col>
      </AxFlexBox.Row>
    </AxFlexBox>
  </HotKeyWrapper>
);

export const ButtonActions: ComponentStory<AnyObject> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      <AxFlexBox.Col>
        <AxButton.Action
          {...props}
          color="success"
          actionType="success"
          message="Action done!!!"
        >
          Success Action
        </AxButton.Action>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton.Action
          {...props}
          color="danger"
          actionType="danger"
          message="Action done!!!"
        >
          Danger Action
        </AxButton.Action>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton.Confirm
          {...props}
          tooltip={{ content: "Test confirmation", placement: "top" }}
          actionType="danger"
          message="Do you confirm?"
        >
          Confirm Action
        </AxButton.Confirm>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton.Dropdown
          {...props}
          label="Click Me"
          tooltip="Test me"
          onClick={action("menuClick")}
        >
          <AxMenu.Item id="menu1" label="Action 1" icon="mdi mdi-bell" />
          <AxMenu.Item id="menu2" label="Action 2" badge="new" />
          <AxMenu.Group label="Grouped" type="collapsable">
            <AxMenu.Item id="menu3" label="Action 3" />
            <AxMenu.Item id="menu4" label="Action 4" />
            <AxMenu.Item id="menu5" label="Action 5" />
          </AxMenu.Group>
        </AxButton.Dropdown>
      </AxFlexBox.Col>
      <AxFlexBox.Col>
        <AxButton.Dropdown
          {...props}
          label="Click Me"
          onClick={action("menuClick")}
        >
          <AxPanel.Stack>
            <AxPanel panelId="one">
              <AxContent padding="none">
                <AxMenu.Item id="menu1" label="Action 1" icon="mdi mdi-bell" />
                <AxMenu.Item id="menu2" label="Action 2" badge="new" />
                <AxMenu.Item
                  data-panel="two"
                  label="Action Next"
                  hotKey="shift+K"
                />
                <AxMenu.Group label="Grouped" type="floating">
                  <AxMenu.Item id="menu3" label="Action 3" />
                  <AxMenu.Item id="menu4" label="Action 4" />
                  <AxMenu.Item id="menu5" label="Action 5" />
                </AxMenu.Group>
              </AxContent>
            </AxPanel>
            <AxPanel panelId="two">
              <AxHeader>
                <AxTitle>Menu Next</AxTitle>
              </AxHeader>
              <AxContent padding="none">
                <AxMenu.Item id="next1" label="Action 1" icon="mdi mdi-bell" />
                <AxMenu.Item id="next2" label="Action 2" badge="new" />
                <AxMenu.Item id="next3" label="Action 3" hotKey="shift+K" />
              </AxContent>
            </AxPanel>
          </AxPanel.Stack>
        </AxButton.Dropdown>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const GroupStory: ComponentStory<AnyObject> = (props) => (
  <AxButton.Group>
    <AxButton {...props} icon="mdi mdi-plus-circle" />
    <AxButton {...props}>Open</AxButton>
    <AxButton {...props}>Again Paste</AxButton>
    <AxButton {...props} color="danger">
      Delete
    </AxButton>
  </AxButton.Group>
);

export const ButtonSamples: ComponentStory<AnyObject> = (props) => {
  return (
    <AxFlexBox>
      <AxFlexBox.Row>
        <AxFlexBox.Col width={200}>
          <AxButton {...props}>This button has a very long label</AxButton>
        </AxFlexBox.Col>
        <AxFlexBox.Col width={600}>
          <AxButton {...props} fullWidth icon="mdi mdi-bell">
            Full width button
          </AxButton>
        </AxFlexBox.Col>
      </AxFlexBox.Row>
    </AxFlexBox>
  );
};

export default { title: "AxButton", component: AxButton };
