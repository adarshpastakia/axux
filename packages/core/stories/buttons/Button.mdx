import {
  ArgsTable,
  Canvas,
  Description,
  Primary,
  Story,
  Title,
} from "@storybook/addon-docs";
import {
  Anchor,
  ArgTypes,
  Controls,
  Meta,
  Subheading,
} from "@storybook/blocks";
import { AxButton, AxCollapsable } from "../../src";
import * as ButtonStories from "./Button.stories";
import * as ButtonActionStories from "./ButtonAction.stories";
import * as ButtonConfirmStories from "./ButtonConfirm.stories";
import * as ButtonDropdownStories from "./ButtonDropdown.stories";
import * as ButtonGroupStories from "./ButtonGroup.stories";

<Meta of={ButtonStories} />

<Anchor storyId="button">
  <div className="invisible h-0 overflow-hidden">
    <Subheading>Button</Subheading>
  </div>
</Anchor>

# Action Buttons

<Description of={AxButton} />

<Primary />

<AxCollapsable>
  <div>Controls</div>
  <Controls of={ButtonStories.Basic} />
</AxCollapsable>

<Anchor storyId="variants">
  <Subheading>Variants</Subheading>
</Anchor>
<Canvas withToolbar of={ButtonStories.Variants} />

<Anchor storyId="colors">
  <Subheading>Colors</Subheading>
</Anchor>
<Canvas withToolbar of={ButtonStories.Colors} />

<Anchor storyId="sizes">
  <Subheading>Sizes</Subheading>
</Anchor>
<Canvas withToolbar of={ButtonStories.Sizes} />

<Anchor storyId="states">
  <Subheading>States</Subheading>
</Anchor>
<Canvas withToolbar of={ButtonStories.States} />

<Anchor storyId="icons">
  <Subheading>Icon Align</Subheading>
</Anchor>
<Canvas withToolbar of={ButtonStories.Icons} />

<Anchor storyId="extras">
  <Subheading>Extras</Subheading>
</Anchor>
<Canvas withToolbar of={ButtonStories.Extras} />

<Anchor storyId="group">
  <Subheading>Group</Subheading>
</Anchor>
<Canvas withToolbar of={ButtonGroupStories.Group} />
<AxCollapsable>
  <div>Controls</div>
  <Controls of={ButtonGroupStories.Group} />
</AxCollapsable>

<Anchor storyId="dropdown">
  <Subheading>Dropdown</Subheading>
</Anchor>
<Canvas withToolbar of={ButtonDropdownStories.Dropdown} />
<AxCollapsable>
  <div>Controls</div>
  <Controls of={ButtonDropdownStories.Dropdown} />
</AxCollapsable>

<Anchor storyId="confirm">
  <Subheading>Confirm</Subheading>
</Anchor>
<Canvas withToolbar of={ButtonConfirmStories.Confirm} />
<AxCollapsable>
  <div>Controls</div>
  <Controls of={ButtonConfirmStories.Confirm} />
</AxCollapsable>

<Anchor storyId="action">
  <Subheading>Action</Subheading>
</Anchor>
<Canvas withToolbar of={ButtonActionStories.Action} />
<AxCollapsable>
  <div>Controls</div>
  <Controls of={ButtonActionStories.Action} />
</AxCollapsable>

<hr />

<Anchor storyId="props">
  <Subheading>Props</Subheading>
</Anchor>

<ArgTypes sort="requiredFirst" />

<Anchor storyId="sample">
  <Subheading>Example</Subheading>
</Anchor>

```tsx
import { AxButton } from "@axux/core";

export const ButtonExample = () => {
  return (
    <AxButton
      badge={{ value: "Value", color: "primary", ping: true }}
      color="primary"
      hotKey="ctrl+enter"
      icon={svgPath || "mdi mdi-icon" || "image.url"}
      iconAlign="start"
      isActive={activeState}
      isDisabled={disabledState}
      isLoading={loadingState}
      nav={routeNav}
      size="md"
      tooltip={{ content: "Tooltip", color: "info", placement: "top" }}
      type="button"
      variant="outline"
      onClick={clickHandler}
      stopPropagation
    >
      Click Me
    </AxButton>
  );
};
```

import { Extras } from "./Button.stories";
