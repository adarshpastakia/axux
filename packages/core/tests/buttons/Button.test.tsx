/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

import { composeStories } from "@storybook/testing-react";
import { fireEvent, render } from "@testing-library/react";
import React from "react";
import * as stories from "../../stories/buttons/Button.stories";

const { Basic } = composeStories(stories);

describe("Button", () => {
  it("should render", () => {
    const fn = jest.fn();
    const fragment = render(<Basic onClick={fn}>Click Me</Basic>);
    expect(
      fragment.container.querySelector(".ax-button__inner"),
    ).not.toBeNull();
    expect(fragment.container.querySelector(".ax-button__icon")).toBeNull();
    expect(
      fragment.container.querySelector(".ax-button__label")?.innerHTML,
    ).toBe("Click Me");
    fireEvent.click(
      fragment.container.querySelector(".ax-button__inner") as HTMLElement,
    );
    expect(fn).toHaveBeenCalled();
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should render icon only", () => {
    const fragment = render(<Basic icon="mdi mdi-bell">{null}</Basic>);
    expect(fragment.container.querySelector(".ax-button__label")).toBeNull();
    expect(fragment.container.querySelector(".ax-button__icon")).not.toBeNull();
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should render icon and label", () => {
    const fragment = render(<Basic icon="mdi mdi-bell">Click Me</Basic>);
    expect(
      fragment.container.querySelector(".ax-button__label"),
    ).not.toBeNull();
    expect(fragment.container.querySelector(".ax-button__icon")).not.toBeNull();
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should be disabled", () => {
    const fn = jest.fn();
    const fragment = render(
      <Basic isDisabled onClick={fn}>
        Click Me
      </Basic>,
    );
    expect(
      fragment.container.querySelector('.ax-button[data-disabled="true"]'),
    ).not.toBeNull();
    fireEvent.click(
      fragment.container.querySelector(".ax-button__inner") as HTMLElement,
    );
    expect(fn).not.toHaveBeenCalled();
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should show loading overlay", () => {
    const fn = jest.fn();
    const fragment = render(
      <Basic isLoading onClick={fn}>
        Click Me
      </Basic>,
    );
    expect(
      fragment.container.querySelector('.ax-button[data-loading="true"]'),
    ).not.toBeNull();
    expect(
      fragment.container.querySelector(".ax-button__loader"),
    ).not.toBeNull();
    fireEvent.click(
      fragment.container.querySelector(".ax-button__inner") as HTMLElement,
    );
    expect(fn).not.toHaveBeenCalled();
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });
});
