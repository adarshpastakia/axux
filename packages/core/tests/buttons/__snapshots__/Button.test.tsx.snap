// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should be disabled 1`] = `"<div></div><div></div><div></div><div><div class="contents" data-color-scheme="light" data-primary-scheme="denim" data-accent-scheme="jade"><div data-active="false" data-round="false" data-square="false" data-spinner="false" data-loading="false" data-disabled="true" class="ax-button "><button role="button" class="ax-button__inner" data-disabled="true" tabindex="0" type="button"><div class="ax-button__label">Click Me</div></button></div><div class="ax-overlay__container" data-mode="overlay"></div><div class="ax-overlay__container" data-mode="alert"></div><div class="ax-notification__container" data-mode="toast"></div><div class="ax-notification__container" data-mode="message"></div></div></div>"`;

exports[`Button should render 1`] = `"<div><div class="contents" data-color-scheme="light" data-primary-scheme="denim" data-accent-scheme="jade"><div data-active="false" data-round="false" data-square="false" data-spinner="false" data-loading="false" data-disabled="false" class="ax-button "><button role="button" class="ax-button__inner" data-disabled="false" tabindex="0" type="button"><div class="ax-button__label">Click Me</div></button></div><div class="ax-overlay__container" data-mode="overlay"></div><div class="ax-overlay__container" data-mode="alert"></div><div class="ax-notification__container" data-mode="toast"></div><div class="ax-notification__container" data-mode="message"></div></div></div>"`;

exports[`Button should render icon and label 1`] = `"<div></div><div></div><div><div class="contents" data-color-scheme="light" data-primary-scheme="denim" data-accent-scheme="jade"><div data-active="false" data-round="false" data-square="false" data-spinner="false" data-loading="false" data-disabled="false" class="ax-button "><button role="button" class="ax-button__inner" data-disabled="false" tabindex="0" type="button"><span role="presentation" class="ax-icon ax-button__icon"><i class="mdi mdi-bell"></i></span><div class="ax-button__label">Click Me</div></button></div><div class="ax-overlay__container" data-mode="overlay"></div><div class="ax-overlay__container" data-mode="alert"></div><div class="ax-notification__container" data-mode="toast"></div><div class="ax-notification__container" data-mode="message"></div></div></div>"`;

exports[`Button should render icon only 1`] = `"<div></div><div><div class="contents" data-color-scheme="light" data-primary-scheme="denim" data-accent-scheme="jade"><div data-active="false" data-round="false" data-square="true" data-spinner="false" data-loading="false" data-disabled="false" class="ax-button "><button role="button" class="ax-button__inner" data-disabled="false" tabindex="0" type="button"><span role="presentation" class="ax-icon ax-button__icon"><i class="mdi mdi-bell"></i></span></button></div><div class="ax-overlay__container" data-mode="overlay"></div><div class="ax-overlay__container" data-mode="alert"></div><div class="ax-notification__container" data-mode="toast"></div><div class="ax-notification__container" data-mode="message"></div></div></div>"`;

exports[`Button should show loading overlay 1`] = `"<div></div><div></div><div></div><div></div><div><div class="contents" data-color-scheme="light" data-primary-scheme="denim" data-accent-scheme="jade"><div data-active="false" data-round="false" data-square="false" data-spinner="false" data-loading="true" data-disabled="false" class="ax-button "><div class="ax-button__loader"></div><button role="button" class="ax-button__inner" data-disabled="true" tabindex="0" type="button"><div class="ax-button__label">Click Me</div></button></div><div class="ax-overlay__container" data-mode="overlay"></div><div class="ax-overlay__container" data-mode="alert"></div><div class="ax-notification__container" data-mode="toast"></div><div class="ax-notification__container" data-mode="message"></div></div></div>"`;
